### ISSUE TITLE
[data grid] Reverse select all behavior

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-x#7753]
### Steps to reproduce 🕹

Link to live example:

Steps:

1. Open https://mui.com/x/react-data-grid/#mit-license-free-forever
2. Click on the checkbox of one row
3. Click on the checkbox of the header

### Current behavior 😯

Unselect all

https://user-images.githubusercontent.com/3165635/215325205-83b1dabd-c361-4b11-a0c5-9190567ac42f.mov

### Expected behavior 🤔

Select all

https://user-images.githubusercontent.com/3165635/215325318-c43bb670-2460-4a9e-9a54-4c896a2c694f.mov

### Context 🔦

You can reproduce the same behavior on:

* [https://airtable.com/](https://airtable.com/)
* [https://www.ag-grid.com/example](https://www.ag-grid.com/example)
* [https://grid.glideapps.com/](https://grid.glideapps.com/)
* [https://stripe.com/](https://stripe.com/)
* [https://tanstack.com/table/v8/docs/examples/react/pagination](https://tanstack.com/table/v8/docs/examples/react/pagination)
* [https://polaris.shopify.com/components/index-table](https://polaris.shopify.com/components/index-table)

I can't find one data grid that behaves like us cc @gerdadesign. I have found this with [https://mui.zendesk.com/agent/tickets/6769](https://mui.zendesk.com/agent/tickets/6769). This is likely creating confusion with end-users. At least, it does for me, I didn't realize it. Subcontionously as an end-user, once you have selected on row, you would try to double click to unselect everything, but with MUI's you would get the opposite of what you want.

### Your environment 🌎

v6.0.0-beta.1

### Order ID 💳 (optional)

*No response*

**Comment by MBilalShafi** [comment]
I agree, it's the same behavior everywhere, even the GitHub issues/PR pages have the same UX.

**Comment by oliviertassinari** [comment]
OK, so it's even something that the community could work on, it's on the MIT side.

**Comment by oliviertassinari** [comment]
Good news, we are consistent with https://mui.com/material-ui/react-checkbox/#indeterminate

**Comment by gerdadesign** [comment]
Yes, agreed!

**Comment by oliviertassinari** [comment]
@DanailH Maybe it could be an onboarding ticket for @noraleonte? The scope is finite/clear/small and it's about design executing.

**Comment by DanailH** [comment]
That's a good suggestion @oliviertassinari but It will be better if she focuses on the pickers mainly because she is part of their team but also because her mentor is also within the team.

**Comment by cherniavskii** [comment]
Interesting, I find the current behavior intuitive 😅

> I can't find one data grid that behaves like us

Gmail is one example:

https://github.com/mui/mui-x/assets/13808724/658f7153-bb99-447b-a45b-4a376b2831c1

But then Google Drive's behavior is the opposite 😅

https://github.com/mui/mui-x/assets/13808724/ccf2fc4c-656d-4a04-a344-dd4fb28b5eb9

**Comment by cherniavskii** [comment]
Related issue: https://github.com/mui/mui-x/issues/3309

**Comment by cndonovan** [comment]
Seconding this issue; my team/my company would love to have the ability to enable "when clicking on header checkbox in an indeterminate state, select all (instead of deselect all)"

**Comment by karlhorky** [comment]
Now that PR #14300 landed in `@mui/x-data-grid-premium@8`, changing the default behavior from "Unselect all" to "Select all" when only partial rows are selected, should this issue be closed?

- #14300

**Comment by karlhorky** [comment]
Also, would it be possible to get a short note about this behavior change with the keyword "Unselect all" in [the Data Grid "Migration from v7 to v8" guide](https://mui.com/x/migration/migration-data-grid-v7/)?
